src/core/backend_context.cc

```diff
@@ -291,24 +295,55 @@ BackendResponder::Finalize()
                 status),
             "error sending TensorFlow response");
       } else {
-        bool cuda_used = false;
-        status = CopyBuffer(
-            response_output->Name(), pinned_memory_type, pinned_memory_id,
-            response_memory_type, response_memory_type_id, response_byte_size,
-            pinned_buffer + offset, const_cast<void*>(response_buffer), stream_,
-            &cuda_used);
-        need_sync_ |= cuda_used;
+        // Only use parallel CopyBuffer for CPU-CPU copy when worker count > 1
+        if ((AsyncWorkQueue::GetWorkerCount() > 1) &&
+            (response_memory_type == TRITONSERVER_MEMORY_CPU) &&
+            (pinned_memory_type == TRITONSERVER_MEMORY_CPU)) {
+          AsyncWorkQueue::AddTask(std::bind(
+              CopyBufferHandler, response_output->Name(), pinned_memory_type,
+              pinned_memory_id, response_memory_type, response_memory_type_id,
+              response_byte_size, pinned_buffer + offset,
+              const_cast<void*>(response_buffer), stream_,
+              &cuda_used_vec[count], &completion_queue));
+          count++;
+        } else {
+          status = CopyBuffer(
+              response_output->Name(), pinned_memory_type, pinned_memory_id,
+              response_memory_type, response_memory_type_id, response_byte_size,
+              pinned_buffer + offset, const_cast<void*>(response_buffer),
+              stream_, &cuda_used_vec[count]);
 
-        if (!status.IsOk()) {
+          if (!status.IsOk()) {
+            LOG_STATUS_ERROR(
+                InferenceResponse::SendWithStatus(
+                    std::move(*response), TRITONSERVER_RESPONSE_COMPLETE_FINAL,
+                    status),
+                "error sending TensorFlow response");
+          }
+        }
+      }
+
+      offset += response_byte_size;
+    }
+
+    for (size_t i = 0; i < cuda_used_vec.size(); i++) {
+      need_sync_ |= cuda_used_vec[i];
+    }
+
+    size_t index = 0;
+    for (auto& pr : def.responses_) {
```

Comment:
You just need to Get from the completion queue 'async_task_count' times and aggregate the status and cuda_used. You can't send response inside the loop as that could result in sending multiple responses if there are multiple errors.

Reply:
Ok. I will rework to return status once. But we should send the response for the first case that fails and break of the loop out right?

Reply:
No, you can't break out of the loop because then you would leave completed task results in the queue. You need to remember on of the failed statuses and then outside the loop return that status.